packages/context/src/index.ts

```diff
@@ -0,0 +1,37 @@
+import { Component, createContext, createComponent, useContext } from "solid-js";
+
+/**
+ * Create the context provider component & useContext function with types inferred from the factory function.
+ * @param factoryFn Factory function will run when the provider component in executed. It takes the provider component `props` as it's argument, and what it returns will be available in the contexts for all the underlying components.
+ * @returns tuple of `[provider component, useContext function]`
+ * @example
+ * ```tsx
+ * const [CounterProvider, useCounter] = createContextProvider((props: { initial: number }) => {
+ *    const [count, setCount] = createSignal(props.initial);
+ *    const increment = () => setCount(count() + 1)
+ *    return { count, increment };
+ * });
+ * // Provide the context
+ * <CounterProvider initial={1}>
+ *    <App/>
+ * </CounterProvider>
+ * // get the context
+ * const ctx = useCounter()
+ * ctx?.count() // => 1
+ * ```
+ */
+export const createContextProvider = <T, P extends Record<string, any>>(
+  factoryFn: (props: P) => T
+): [provider: Component<P>, useContext: () => T | undefined] => {
```

Comment:
`=> T | undefined` 
Maybe remove that undefined assignment?

Reply:
Will be adding a second argument for the defaults, if you provide them the type will switch to `T`. If you want a definite type without defaults, you're breaking the law, but I guess you could do it like this: *(probably should add this to the readme...)*
```ts
const useDefiniteCounter = () => useCounter()!
```

Reply:
Maybe the default argument should be of generic type `D`, and the resulting type is `T | D`? This would work even when the default argument is undefined, and allows providing some of the interface but not all in the default behavior. (I've wanted this in the past, though I might be weird.)

Reply:
I sometimes do the same in similar fallback-needing apis, but the worst part of this solution is that you loose the type suggestions/warnings. So I would leave that to the user to implement, for example:
```ts
const useCounterOrSomethingElse = () => useCounter() ?? fallback_with_different_type
```

Reply:
I sometimes do the same in similar fallback-needing apis, but the worst part of this solution is that you loose the type suggestions/warnings. So I would leave that to the user to implement, for example:
```ts
const useCounterOrSomethingElse = () => useCounter() ?? fallback_with_different_type
```